src/Bedrock.Framework/Protocols/ConsumableArrayBufferWriter.cs

```diff
@@ -0,0 +1,235 @@
+﻿using System;
+using System.Buffers;
+using System.Diagnostics;
+
+namespace Bedrock.Framework.Protocols
+{
+    /// <summary>
+    /// This should be replaced with the framework version if https://github.com/dotnet/runtime/issues/1248 is implemented
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class ConsumableArrayBufferWriter<T> : IBufferWriter<T>
+    {
+        private T[] _buffer;
+        private int _index;
+        private int _consumedCount;
+
+        private const int DefaultInitialBufferSize = 256;
+
+        /// <summary>
+        /// Creates an instance of an <see cref="ConsumableArrayBufferWriter{T}"/>, in which data can be written to,
+        /// with the default initial capacity.
+        /// </summary>
+        public ConsumableArrayBufferWriter()
+        {
+            _buffer = Array.Empty<T>();
+            _index = 0;
+            _consumedCount = 0;
+        }
+
+        /// <summary>
+        /// Creates an instance of an <see cref="ConsumableArrayBufferWriter{T}"/>, in which data can be written to,
+        /// with an initial capacity specified.
+        /// </summary>
+        /// <param name="initialCapacity">The minimum capacity with which to initialize the underlying buffer.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <paramref name="initialCapacity"/> is not positive (i.e. less than or equal to 0).
+        /// </exception>
+        public ConsumableArrayBufferWriter(int initialCapacity)
+        {
+            if (initialCapacity <= 0)
+                throw new ArgumentException(nameof(initialCapacity));
+
+            _buffer = new T[initialCapacity];
+            _index = 0;
+            _consumedCount = 0;
+        }
+
+        /// <summary>
+        /// Returns the unconsumed data written to the underlying buffer so far, as a <see cref="ReadOnlyMemory{T}"/>.
+        /// </summary>
+        public ReadOnlyMemory<T> WrittenMemory => _buffer.AsMemory(_consumedCount, _index - _consumedCount);
```

Comment:
```suggestion
        public ReadOnlyMemory<T> WrittenMemory => _buffer.AsMemory(_consumedCount.._index);
```
?
Same below for the span.